### PR TITLE
Add learnable XSA

### PR DESCRIPTION
**Pull request by CerovazS**
# Add learnable XSA

## Summary

This PR adds learnable [Exclusive Self-Attention (XSA)](https://arxiv.org/pdf/2603.09078) to `tiny/train.py`, applied by default to the first 6 transformer layers.

The default candidate setting is:

```bash
torchrun --standalone --nproc_per_node=8 tiny/train.py \
  --xsa-mode first6 \
  --max-train-steps 3040
```

The original baseline path remains available with `--xsa-mode off`.

## What changed

- Add learnable XSA for the first 6 transformer layers, controlled by:
  - `--xsa-mode off`
  - `--xsa-mode first6`
- Make `--xsa-mode first6` the default, while preserving `--xsa-mode off` for the baseline path.
- Add a fixed short-run step budget:
  - `--max-train-steps`
- Make `--max-train-steps 3040` the default.

## Main result

Baseline references:

| Run | Ckpt avg val loss | Total training time |
| --- | ---: | ---: |
| Reproduced Tiny baseline | `3.332274` | `14.76m` |
| Official Tiny record attached log | `3.332247` | `14.70m` |

XSA first-6 result:

| Run | Setting | Ckpt avg val loss | Total training time |
| --- | --- | ---: | ---: |
| `tiny_xsa_first6_8xh100_3040_20260602T111128Z` | `--xsa-mode first6 --max-train-steps 3040` | `3.329040` | `14.71m` |

This XSA first-6 run is under the 15-minute training-time budget and beats the reproduced baseline and official attached-log score on checkpoint-averaged validation loss.

## Delta

XSA first-6 run:

- vs reproduced baseline: `3.329040 - 3.332274 = -0.003234`
- vs official attached-log score: `3.329040 - 3.332247 = -0.003207`

## Epoch validation losses

Run: `tiny_xsa_first6_8xh100_3040_20260602T111128Z`

| Epoch | Step | Val loss |
| ---: | ---: | ---: |
| 1 | 190 | `4.488313` |
| 2 | 380 | `4.076740` |
| 3 | 570 | `3.947173` |
| 4 | 760 | `3.863204` |
| 5 | 950 | `3.783885` |
| 6 | 1140 | `3.712024` |
| 7 | 1330 | `3.633137` |
| 8 | 1520 | `3.540873` |
| 9 | 1710 | `3.522695` |
| 10 | 1900 | `3.541720` |
| 11 | 2090 | `3.546714` |
| 12 | 2280 | `3.544263` |
| 13 | 2470 | `3.371458` |
| 14 | 2660 | `3.367202` |
| 15 | 2850 | `3.369194` |
| 16 | 3040 | `3.393009` |

Final metrics:

- Final cutoff validation loss: `3.393009`
- EMA validation loss: `3.346719`
- Checkpoint-averaged validation loss: `3.329040`
- Wall clock time: `19.27m`
- Total training time: `14.71m`
- Peak memory: `70,182 MiB`

## Acknowledgements

This work was also possible thanks to the support provided by Paradigma ([link](https://paradigma.inc/)) and the use of Flywheel ([link](https://flywheel.paradigma.inc/)), Paradigma's infrastructure for research.

Team: @CerovazS, @MarioPaerle, @GabrieleCirillo, @Mister-dev-oss, @crisostomi.

### CODE REVIEW

**Comment by akshayvegesna** [comment]
Thanks! Can you make these flags default `--xsa-mode first6 --max-train-steps 3040`, and also remove all the new code unrelated from these flags?

**Comment by CerovazS** [comment]
Done, thanks!
I also updated the PR description to remove references to the removed experimental flags.
